### ISSUE TITLE
Invariant Proof: Fixed Exchange Rate

### ISSUE THREAD

**Issue by lucasmt** [golemfoundation/octant-v2-core#61]
## Invariant Statement

**Invariant:** `totalSupply == totalAssets`, unless the strategy suffers a loss greater than can be recouped by burning shares of the `DragonRouter`.

## Auxiliary Lemma: Conversion between Shares and Assets

Before showing that the invariant holds, we prove the following auxiliary result:

* If `totalSupply == totalAssets`, then `_convertToShares(S, assets, rounding) == assets` and `_convertToAssets(S, shares, rounding) == shares`.

The implementation of `_convertToShares` has three cases:

1. If `totalSupply == 0`, then it returns `assets` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L672)).
2. Otherwise, if `totalAssets == 0`, then it returns `0` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L676)). However, this case cannot happen if `totalSupply == totalAssets`, since we would have returned in case (1).
3. Otherwise, it uses the OpenZeppelin library function `mulDiv` to return `(assets * totalSupply) / totalAssets`, rounding up or down depending on the `rounding` parameter ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L678)). In either case, since `totalSupply == totalAssets`, the division is exact and returns `assets` (unless there is an overflow, in which case it would revert).

Therefore, in all valid cases the function returns `assets`.

The implementation of `_convertToAssets` similarly has two cases ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L690)):

1. If `totalSupply == 0`, then it returns `shares`.
2. Otherwise, it uses `mulDiv` to return `(shares * totalAssets) / totalSupply`. Again, since `totalSupply == totalAssets`, the division is exact regardless or rounding and returns `shares` as long as it doesn't revert due to overflow.

Therefore, the function returns `shares` in both cases.

Since the storage data `S` is implied and, as we showed above, `rounding` does not affect the result under the assumption that `totalSupply == totalAssets`, below we use `_convertToShares(assets)` and `_convertToAssets(shares)` as a shorthand.

## Invariant Base Case

When the contract is deployed, `totalSupply == 0` and `totalAssets == 0`, therefore the invariant holds.

## Invariant Induction Step

We assume that the invariant holds in the current state. Then, if a loss that is greater than the shares of the `DragonRouter` hasn't occurred yet, `totalSupply == totalAssets`. We show that the invariant is maintained after every external call; that is, that `totalSupply' == totalAssets'`, where `X'` denotes the value of the variable `X` after the call returns.

The external functions that might affect this invariant are the following:

* The external functions that update `totalAssets` are `report`, those that call `TokenizedStrategy._deposit` (`deposit`, `depositWithLockup`, `mint`, `mintWithLockup`), and those that call `TokenizedStrategy._withdraw` (`withdraw`, `redeem`).
* The external functions that update `totalSupply` are those that call `TokenizedStrategy._mint` (`report`, `deposit`, `depositWithLockup`, `mint`, `mintWithLockup`) and those that call `TokenizedStrategy._burn` (`report`, `withdraw`, `redeem`).

We consider each of these external functions and show that they preserve the invariant.

### `report`

This function harvests the funds and stores the new value in the variable `newTotalAssets` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L446)), then updates `totalAssets` to `newTotalAssets` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L467)):

```
totalAssets' == newTotalAssets
```

As for `totalSupply`, there are two cases:

1. If `newTotalAssets > oldTotalAssets` (where `oldTotalAssets` is the value of `totalAssets` before the update), then it converts `profit = newTotalAssets - oldTotalAssets` to shares and mints that amount of shares to the `DragonRouter` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L454)), increasing the total supply.
2. If `newTotalAssets <= oldTotalAssets`, then it calculates `loss = oldTotalAssets - newTotalAssets` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L457)). If `loss == 0`, then `totalSupply` is not updated. Otherwise ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L460)), `loss` is converted to shares ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L416)) and those shares are burned from the `DragonRouter`, as long as there are enough shares to be burned ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L420)).

In case (1), we have that

```
totalSupply' == totalSupply + _convertToShares(profit)
```

Since we assume that `totalSupply == totalAssets` holds before the update, 
`_convertToShares(profit) == profit`. Therefore,

```
totalSupply' == totalSupply + _convertToShares(profit)
             == totalSupply + profit
             == totalSupply + newTotalAssets - oldTotalAssets
             == totalSupply + newTotalAssets - totalAssets
             == totalAssets + newTotalAssets - totalAssets
             == newTotalAssets
```

Therefore, `totalSupply' == totalAssets'`, and the invariant holds after the call.

In case (2), there are three sub-cases:

1. If `loss == 0`, then `totalSupply' == totalSupply`. But if `loss == 0` this also means `newTotalAssets == oldTotalAssets`, and therefore `totalAssets' == totalAssets`. Therefore, the invariant is preserved.
2. If `loss > 0` and `_convertToShares(loss) > balances[dragonRouter]`, then the invariant is no longer required to hold. Therefore, we can ignore this case.
3. If `loss > 0` and `_convertToShares(loss) <= balances[dragonRouter]`, then `totalSupply` is updated to `totalSupply - _convertToShares(loss)`. Again, since `totalSupply == totalAssets`, we get `_convertToShares(loss) == loss`, and therefore

```
totalSupply' == totalSupply - _convertToShares(loss)
             == totalSupply - loss
             == totalSupply - (oldTotalAssets - newTotalAssets)
             == totalSupply - oldTotalAssets + newTotalAssets
             == totalSupply - totalAssets + newTotalAssets
             == totalAssets - totalAssets + newTotalAssets
             == newTotalAssets
```

Therefore, `totalSupply' == totalAssets'`, and the invariant holds in this case as well.

### `deposit` and `depositWithLockup`

These two functions receive a parameter `assets` and update `totalAssets` and `totalSupply` as follows via `TokenizedStrategy._deposit` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L377)):

```
totalAssets' == totalAssets + assets
totalSupply' == totalSupply + shares
```

where `shares = _convertToShares(assets)` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L364)). Since `totalSupply == totalAssets`, this means `shares == assets`. Therefore,

```
totalAssets' == totalAssets + assets
totalSupply' == totalAssets + assets
```

Therefore, `totalSupply' == totalAssets'`.

### `mint` and `mintWithLockup`

These two functions receive a parameter `shares` and convert it to `assets = _convertToAssets(shares)` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L402)). They then pass `assets` as input to `DragonTokenizedStrategy._deposit` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L406)), which makes the same update as `deposit` and `depositWithLockup`. Therefore, the same reasoning above holds, and `totalSupply' == totalAssets'`.

### `withdraw` and `redeem`

Both of these functions call `TokenizedStrategy._withdraw` with parameters `shares` and `assets`:

* `withdraw` receives `assets` as an argument and calculates `shares = _convertToShares(assets)` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L272)).
* `redeem` receives `shares` as an argument and calculates `assets = _convertToAssets(shares)` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/DragonTokenizedStrategy.sol#L314)).

In either case, assuming `totalSupply == totalAssets`, we have that `shares == _convertToShares(assets) == assets` and `assets == _convertToAssets(shares) == shares`.

`TokenizedStrategy._withdraw` updates `totalShares` by calling the `TokenizedStrategy._burn` function, which subtracts `shares` from the `totalSupply` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L836)):

```
totalSupply' == totalSupply - shares
```

It also updates `totalAssets` as follows ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L834)):

```
totalAssets' == totalAssets - (assets' + loss)
```

where `assets'` and `loss` depend on three cases ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L809)):

1. If `idle >= assets`, where `idle` is the amount of idle funds currently sitting in the contract, then `assets' = assets` and `loss = 0`.
2. Otherwise, it calls `BaseStrategy.freeFunds` to try to free the difference from the underlying strategy ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L812)). If it's successful, then again `assets' = assets` and `loss = 0`.
3. If it's not able to free enough funds, then `loss = assets - idle'`, where `idle'` is the new amount of idle funds after freeing ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L821)), and `assets' = idle'` ([link](https://github.com/golemfoundation/octant-v2-core/blob/f65cac2db7ea53c722916efc060d90c4c6748cc8/src/dragons/vaults/TokenizedStrategy.sol#L829)).

In cases (1) and (2),

```
totalAssets' == totalAssets - (assets' + loss)
             == totalAssets - (assets + 0)
             == totalAssets - assets
```

In case (3),

```
totalAssets' == totalAssets - (assets' + loss)
             == totalAssets - (idle' + assets - idle')
             == totalAssets - assets
```

In either case, since `totalSupply == totalAssets` and `shares == assets`, we have that `totalSupply' == totalAssets'`.
